src/main/323C.c

```diff
@@ -685,7 +685,79 @@ INCLUDE_ASM("asm/us/main/nonmatchings/323C", func_80017100);
 
 INCLUDE_ASM("asm/us/main/nonmatchings/323C", func_80017268);
 
-INCLUDE_ASM("asm/us/main/nonmatchings/323C", func_80017340);
+void func_80017340(void)
+{
+    u8 start;
+    s8 end;
+    u8 var_v0;
+
+    *(void**)0x1F800100 = &D_801499C8[*(s32*)0x1F800000];
+    *(void**)0x1F800104 = &D_80169D98[*(s32*)0x1F800000];
+    *(void**)0x1F800110 = &D_801499C8[*(s32*)0x1F800000];
+
+    func_80017E84();
+
+    if (engine_obj.cur_character == CHARACTER_X) {
+        end = 10;
+    } else {
+        end = 8;
+    }
+
+    if (engine_obj.cur_character == CHARACTER_X) {
+        func_800175AC(0x14);
+        func_800175AC(0);
+        func_800175AC(1);
+        if (bar_object.base.unk17 != 0) {
+            func_800175AC(0x15);
+            func_800175AC(2);
+            func_800175AC(3);
+        }
+    }
+
+    if (engine_obj.cur_character == CHARACTER_X) {
+        start = 2;
+    } else {
+        start = 0;
+    }
+
+    while (start < end) {
+        if (((u8*)&bar_object.base.unk16)[start] != 0) {
+            func_800175AC(start + 0x14);
+            func_800175AC(start * 2);
+            func_800175AC(start * 2 + 1);
+        }
+        start++;
+    }
+
+    if (bar_object.unk20 != 0) {
+        func_800175AC(0x23);
+    }
+    if (bar_object.unk21 != 0) {
+        func_800175AC(0x24);
+    }
+    if (bar_object.unk22 != 0) {
+        func_800175AC(0x25);
+    }
+    func_800175AC(0x1F);
+    if (bar_object.unk25 != 0) {
+        func_800175AC(0x26);
+    }
+    func_800175AC(0x1E);
+    func_800175AC(0x20);
+    if ((u8)engine_obj.stage - 1 < 8U) {
+        if (engine_obj.cur_character == CHARACTER_X) {
+            var_v0 = ((u8*)&bar_object.base.unk17)[engine_obj.stage];
+        } else {
+            var_v0 = ((u8*)&bar_object.base.unk16)[D_800F1C0F[engine_obj.stage]];
```

Comment:
I'm thinking it's like this since these overlap
```
            var_v0 = ((u8*)&bar_object.base.unk16)[engine_obj.stage + 1];
        } else {
            var_v0 = ((u8*)&bar_object.base.unk16)[D_800F1C0F[engine_obj.stage]];
```